Extensions/render/opticl-image.lisp

```diff
@@ -0,0 +1,296 @@
+    (make-fill-image-function
+     (opticl-image-data-get-pixel data-image i j)
+     (opticl-image-data-set-pixel data-image i j red green blue alpha)
+     (values 
+      (uniform-rgba-design-red rgba-design)
+      (uniform-rgba-design-green rgba-design)
+      (uniform-rgba-design-blue rgba-design)
+      (uniform-rgba-design-alpha rgba-design))
+     255))
+  (make-rectangle* x y (+ x width) (+ y height)))
+
+(defmethod fill-image ((image opticl-image) rgba-design (mask (eql nil))
+		       &key
+			 (x 0) (y 0)
+			 (width (climi::image-width image)) (height (climi::image-height image))
+			 (mask-dx 0) (mask-dy 0))
+  (declare (type fixnum x y width height mask-dx mask-dy)
+	   (ignore mask-dx mask-dy))
+  (let ((data-image (image-data image))
+	(source-fn (make-rgba-design-fn rgba-design)))
+    (declare (type opticl-image-data data-image)
+	     (type design-fn source-fn))
+    (make-fill-image-function
+     (opticl-image-data-get-pixel data-image i j)
+     (opticl-image-data-set-pixel data-image i j red green blue alpha)
+     (funcall source-fn i j)
+     255))
+  (make-rectangle* x y (+ x width) (+ y height)))
+
+
+(defmethod fill-image ((image opticl-image) (rgba-design uniform-rgba-design) (mask mask-image)
+		       &key
+			 (x 0) (y 0)
+			 (width (climi::image-width image)) (height (climi::image-height image))
+			 (mask-dx 0) (mask-dy 0))
+  (declare (type fixnum x y width height mask-dx mask-dy))
+  (let ((data-image (image-data image))
+	(data-mask (image-data mask)))
+    (declare (type opticl-image-data data-image)
+	     (type mask-image-data data-mask))
+    (make-fill-image-function
+     (opticl-image-data-get-pixel data-image i j)
+     (opticl-image-data-set-pixel data-image i j red green blue alpha)
+     (values 
+      (uniform-rgba-design-red rgba-design)
+      (uniform-rgba-design-green rgba-design)
+      (uniform-rgba-design-blue rgba-design)
+      (uniform-rgba-design-alpha rgba-design))
+     (mask-image-data-get-alpha data-mask (+ mask-dx i) (+ mask-dy j))))
+  (make-rectangle* x y (+ x width) (+ y height)))
+  
+(defmethod fill-image ((image opticl-image) rgba-design (mask mask-image)
+		       &key
+			 (x 0) (y 0)
+			 (width (climi::image-width image)) (height (climi::image-height image))
+			 (mask-dx 0) (mask-dy 0))
+  (declare (type fixnum x y width height mask-dx mask-dy))
+  (let ((data-image (image-data image))
+	(data-mask (image-data mask))
+	(source-fn (make-rgba-design-fn rgba-design)))
+    (declare (type opticl-image-data data-image)
+	     (type mask-image-data data-mask)
+	     (type design-fn source-fn))
+    (make-fill-image-function
+     (opticl-image-data-get-pixel data-image i j)
+     (opticl-image-data-set-pixel data-image i j red green blue alpha)
+     (funcall source-fn i j)
+     (mask-image-data-get-alpha data-mask (+ mask-dx i) (+ mask-dy j))))
+  (make-rectangle* x y (+ x width) (+ y height)))
+
+;;;
+;;;
+;;;
+
+;;;
+;;; macro to build drawing function
+;;;
+
+
+
+
+
```

Comment:
unjustified empty space. If you want to spearate this parts more visibly, you may add page break, or a comment box